polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -46,17 +47,48 @@ namespace {
 // Utilities functions
 //===----------------------------------------------------------------------===//
 
-/// A function is a candidate iff is a kernel body functions with an nd_item
-/// argument.
-bool isCandidateFunction(FunctionOpInterface func) {
-  if (!polygeist::isPotentialKernelBodyFunc(func))
+bool isLocalAccessAddrSpace(Type ty) {
+  if (auto memRefTy = dyn_cast<MemRefType>(ty)) {
+    if (auto memSpace = dyn_cast_or_null<sycl::AccessAddrSpaceAttr>(
+            memRefTy.getMemorySpace())) {
+      if (memSpace.getValue() == sycl::AccessAddrSpace::LocalAccess)
+        return true;
+      return false;
+    }
+    return (memRefTy.getMemorySpaceAsInt() == 3);
+  }
+  if (auto ptrTy = dyn_cast<LLVM::LLVMPointerType>(ty))
+    return (ptrTy.getAddressSpace() == 3);
+  return false;
+}
+
+/// A kernel is a candidate iff no dynamic sized local accessor is used.
+bool isCandidateKernel(gpu::GPUFuncOp kernel) {
```

Comment:
Technically a `gpu::GPUFuncOp` is not a kernel unless it has the 'kernel' attribute. Can you check it has that attribute and return false if it doesn't (or assert).